packages/fetchai/connections/http/connection.py

```diff
@@ -0,0 +1,359 @@
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+
+"""HTTP connection, channel, and server"""
+
+import asyncio
+import logging
+import threading
+import time
+import yaml
+import json
+from asyncio import CancelledError
+from threading import Thread
+from typing import Any, Dict, List, Optional, Set, cast
+from http.server import HTTPServer, BaseHTTPRequestHandler
+
+from aea.configurations.base import ConnectionConfig, PublicId
+from aea.connections.base import Connection
+from aea.mail.base import Address, Envelope, EnvelopeContext, URI
+
+
+logger = logging.getLogger(__name__)
+
+
+class HTTPHandler(BaseHTTPRequestHandler):
+
+    def do_HEAD(s):
+        # Placeholder functions
+        s.send_response(200)
+        s.send_header("Content-type", "text/html")
+        s.end_headers()
+
+    def do_GET(s):
+        """Respond to a GET request."""
+        # Placeholder functions
+        s.send_response(200)
+        s.send_header("Content-type", "text/html")
+        s.end_headers()
+        s.wfile.write("<html><head><title>Title goes here.</title></head>")
+        s.wfile.write("<body><p>This is a test.</p>")
+        # If someone went to "http://something.somewhere.net/foo/bar/",
+        # then s.path equals "/foo/bar/".
+        s.wfile.write("<p>You accessed path: %s</p>" % s.path)
+        s.wfile.write("</body></html>")
+
+    def do_POST(s):
+        """Respond to a GET request."""
+        pass
+
+
+class HTTPChannel:
+    """A wrapper for an RESTful API with an internal HTTPServer."""
+
+    def __init__(self,
+                 # address: Address,
+                 api_spec: str,
+                 host: str,
+                 port: int,
+                 excluded_protocols: Optional[Set[PublicId]] = None,
+                 ):
+        """
+        Initialize a channel.
+
+        :param address: the address
+        """
+        # self.address = address
+        self.host = host
+        self.port = port
+        # self.uri_base = f"http://{host}:{port}"
+        self.in_queue = None  # type: Optional[asyncio.Queue]
+        self.loop = None  # type: Optional[asyncio.AbstractEventLoop]
+        self.excluded_protocols = excluded_protocols
+        self.thread = Thread(target=self.receiving_loop)
+        self.lock = threading.Lock()
+        self.stopped = True
+        # Initializing the internal HTTP server
+        try:
+            self.server = HTTPServer((self.host, self.port), HTTPHandler)
+        except OSError:
+            logger.error(f"{host}:{port} is already in use, please try another Socket.")
+        # Process API specifications into Envelopes to be sent to InBox
+        self._process_api(api_spec)
```

Comment:
I would use the api_spec to filter for valid paths which the handlers should process